metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+              tokenDefaultMaxLifetime,
+              tokenDefaultRenewLifetime);
+        }
+    }
+
+    private final Logger log;
+    private final DelegationTokenCache tokenCache;
+    private final String secretKeyString;
+    private final long tokenDefaultMaxLifetime;
+    private final long tokenDefaultRenewLifetime;
+
+    private DelegationTokenControlManager(
+        LogContext logContext,
+        DelegationTokenCache tokenCache,
+        String secretKeyString,
+        long tokenDefaultMaxLifetime,
+        long tokenDefaultRenewLifetime
+    ) {
+        this.log = logContext.logger(DelegationTokenControlManager.class);
+        this.tokenCache = tokenCache;
+        this.secretKeyString = secretKeyString;
+        this.tokenDefaultMaxLifetime = tokenDefaultMaxLifetime;
+        this.tokenDefaultRenewLifetime = tokenDefaultRenewLifetime;
+    }
+
+    public static byte[] toBytes(String str) {
+        return str.getBytes(StandardCharsets.UTF_8);
+    }
+
+    private byte[] createHmac(String tokenId) throws Exception {
+        Mac mac = Mac.getInstance("HmacSHA512");
+        SecretKeySpec secretKey = new SecretKeySpec(toBytes(secretKeyString), mac.getAlgorithm());
+
+        mac.init(secretKey);
+        return mac.doFinal(toBytes(tokenId));
+    }
+
+    private TokenInformation getToken(byte[] hmac) {
+        String base64Pwd = Base64.getEncoder().encodeToString(hmac);
+        return tokenCache.tokenForHmac(base64Pwd);
+    }
+
+    private boolean allowedToRenew(TokenInformation tokenInfo, KafkaPrincipal renewer) {
+        if (tokenInfo.owner().equals(renewer)) {
+            return true;
+        }
+        for (KafkaPrincipal validRenewer : tokenInfo.renewers()) {
+            if (validRenewer.equals(renewer)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean isEnabled() {
+        if (secretKeyString != null) {
+            return true;
+        }
+        return false;
+    }
+
+    /*
+     * Pass in the MetadataVersion so that we can return a response to the caller 
+     * if the current metadataVersion is too low.
+     */
+    public ControllerResult<CreateDelegationTokenResponseData> createDelegationToken(
+        ControllerRequestContext context,
+        CreateDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+        long maxLifeTime = tokenDefaultMaxLifetime;
+        if (requestData.maxLifetimeMs() > 0) {
+            maxLifeTime = Math.min(maxLifeTime, requestData.maxLifetimeMs());
+        }
+
+        long maxTimestamp = now + maxLifeTime;
+        long expiryTimestamp = Math.min(maxTimestamp, now + tokenDefaultRenewLifetime);
+
+        String tokenId = Uuid.randomUuid().toString();
+
+        KafkaPrincipal owner = context.principal();
+        if ((requestData.ownerPrincipalName() != null) && 
+            (!requestData.ownerPrincipalName().isEmpty())) {
```

Comment:
Any particular reason for all the extra parens?

Reply:
Readability. It calls to attention the != null and the !.isEmpty()